### ISSUE TITLE
Assertion `buf->b_signcols.count[width - 1] >= 0` failed

### ISSUE THREAD

**Issue by przepompownia** [neovim/neovim#27046]
### Problem

From few days I get crashes when updating the buffer e.g. undo. 

```cpp
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007fea9eef315f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007fea9eea5472 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fea9ee8f4b2 in __GI_abort () at ./stdlib/abort.c:79
#4  0x00007fea9ee8f3d5 in __assert_fail_base (fmt=0x7fea9f003dc8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x5565767fe960 "buf->b_signcols.count[width - 1] >= 0",
    file=file@entry=0x5565767fe748 ".../src/nvim/decoration.c", line=line@entry=848, function=function@entry=0x55657679f310 <__PRETTY_FUNCTION__.1.lto_priv.8> "buf_signcols_count_range") at ./assert/assert.c:92
#5  0x00007fea9ee9e3a2 in __assert_fail (assertion=assertion@entry=0x5565767fe960 "buf->b_signcols.count[width - 1] >= 0", file=file@entry=0x5565767fe748 ".../src/nvim/decoration.c", line=line@entry=848,
    function=function@entry=0x55657679f310 <__PRETTY_FUNCTION__.1.lto_priv.8> "buf_signcols_count_range") at ./assert/assert.c:101
#6  0x00005565764923be in buf_signcols_count_range (buf=0x55657a5cb7f0, row1=20, row2=20, add=-1, clear=kFalse) at .../src/nvim/decoration.c:848
#7  0x000055657648f9c0 in buf_remove_decor_sh (buf=buf@entry=0x55657a5cb7f0, row1=row1@entry=20, row2=row2@entry=20, sh=sh@entry=0x556578824ac0) at .../src/nvim/decoration.c:258
#8  0x000055657648fa46 in buf_decor_remove (buf=0x55657a5cb7f0, row1=20, row2=20, decor=..., free=true) at .../src/nvim/decoration.c:227
#9  0x0000556576516e98 in extmark_del (buf=buf@entry=0x55657a5cb7f0, itr=itr@entry=0x7fff47f5d040, key=..., restore=restore@entry=true) at .../src/nvim/extmark.c:153
#10 0x0000556576517558 in extmark_clear (buf=0x55657a5cb7f0, ns_id=53, l_row=0, l_col=<optimized out>, u_row=2147483646, u_col=2147483647) at .../src/nvim/extmark.c:189
#11 0x0000556576444f04 in nvim_buf_clear_namespace (buffer=<optimized out>, ns_id=53, line_start=0, line_end=2147483647, err=0x7fff47f5d190) at .../src/nvim/api/extmark.c:973
#12 0x0000556576429c02 in nlua_api_nvim_buf_clear_namespace (lstate=0x7fea8a4ac5f0) at .../build/src/nvim/auto/lua_api_c_bindings.generated.c:2922
#13 0x0000556576726ed6 in lj_BC_FUNCC ()
#14 0x0000556576727d32 in lj_ff_coroutine_resume ()
#15 0x00005565767132a5 in lua_pcall (L=L@entry=0x7fea9ee47380, nargs=nargs@entry=0, nresults=nresults@entry=0, errfunc=errfunc@entry=-2) at lj_api.c:1121
#16 0x000055657654d077 in nlua_pcall (lstate=lstate@entry=0x7fea9ee47380, nargs=nargs@entry=0, nresults=nresults@entry=0) at .../src/nvim/lua/executor.c:172
#17 0x0000556576550cbc in nlua_schedule_event (argv=<optimized out>) at .../src/nvim/lua/executor.c:367
#18 0x0000556576639898 in state_handle_k_event () at .../src/nvim/state.c:119
#19 0x00005565765a1c99 in nv_event (cap=0x7fff47f5d460) at .../src/nvim/normal.c:6613
#20 0x000055657659d9d1 in normal_execute (state=0x7fff47f5d3f0, key=<optimized out>) at .../src/nvim/normal.c:1227
#21 0x000055657663973b in state_enter (s=0x7fff47f5d3f0) at .../src/nvim/state.c:101
#22 0x000055657659735d in normal_enter (cmdwin=<optimized out>, noexmode=<optimized out>) at .../src/nvim/normal.c:516
#23 0x00005565765601bd in main (argc=2, argv=<optimized out>) at .../src/nvim/main.c:651
```

@luukvbaal it looks like from https://github.com/neovim/neovim/commit/967c7abde3c6fa3210a4920a5848a54dc913d851#diff-6e7593afe2030cd8628b9a5a900e49accd4a9d7f3ff4b73ee022f06b4bd0f237R848

### Steps to reproduce

At the moment I have no idea how to reproduce this bug. I use statuscol nvim on branch 0.10 with 
```lua
local builtin = require('statuscol.builtin')
require('statuscol').setup({
  relculright = true,
  segments = {
    {
      text = {builtin.foldfunc},
      condition = {true},
      click = 'v:lua.ScFa',
    },
    {
      sign = {namespace = {'gitsigns*'}, maxwidth = 1, colwidth = 1, auto = true},
      click = 'v:lua.ScSa',
    },
    {
      sign = {namespace = {'arctgx.diagnostic.signs/diagnostic/signs'}, maxwidth = 2, colwidth = 2, auto = true},
      click = 'v:lua.ScSa'
    },
    {
      sign = {name = {'Dap.*'}, maxwidth = 2, colwidth = 2, auto = true},
      click = 'v:lua.ScLa'
    },
    {
      sign = {name = {'.*'}, maxwidth = 1, colwidth = 1, auto = true},
      click = 'v:lua.ScSa'
    },
    {text = {builtin.lnumfunc, ' '}, click = 'v:lua.ScLa',},
  },
  ft_ignore = {
    'NvimTree',
  },
})
```

### Expected behavior

No crash

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2104+gfd2ed024c

### Vim (not Nvim) behaves the same?

-

### Operating system/version

Debian Sid

### Terminal name/version

alacritty 0.13.1

### $TERM environment variable

alacritty

### Installation

from repo

**Comment by wookayin** [comment]
Possibly a side effect of #26898

**Comment by przepompownia** [comment]
Config for diagnostics include unicode chars:
```lua
vim.diagnostic.config({
  virtual_text = false,
  underline = false,
  float = {
    border = 'rounded',
  },
  signs = {
    text = {
      [vim.diagnostic.severity.ERROR] = '',
      [vim.diagnostic.severity.WARN] = '⚠',
      [vim.diagnostic.severity.INFO] = '',
      [vim.diagnostic.severity.HINT] = '󰌶',
    },
  },
})

local ns = vim.api.nvim_create_namespace('arctgx.diagnostic.signs')
local origSignsHandler = vim.diagnostic.handlers.signs

-- from :help diagnostic-handlers-example
vim.diagnostic.handlers.signs = {
  show = function (_, bufnr, _, opts)
    local diagnostics = vim.diagnostic.get(bufnr)

    local maxSeverityPerLine = {}
    for _, d in pairs(diagnostics) do
      local m = maxSeverityPerLine[d.lnum]
      if not m or d.severity < m.severity then
        maxSeverityPerLine[d.lnum] = d
      end
    end

    local filteredDiagnostics = vim.tbl_values(maxSeverityPerLine)
    origSignsHandler.show(ns, bufnr, filteredDiagnostics, opts)
  end,
  hide = function (_, bufnr)
    if vim.api.nvim_buf_is_valid(bufnr) then
      origSignsHandler.hide(ns, bufnr)
    end
  end,
}
```

**Comment by przepompownia** [comment]
> Possibly a side effect of #26898

yes, I quoted it above too.

**Comment by luukvbaal** [comment]
A repro would be nice. The undo/redo part is tricky but I haven't hit the assert myself. 

I think it is worthwhile to keep trying to make the approach in #26898 work. The fact that `b_signcols.count` is decremented below 0 indicates an error in keeping track of the number of lines with a certain number of signs. In the meantime we could merge #27047, that doesn't actually fix the problem but `abort()'`ing on error might be unnecessary(even if it is on Debug builds).

**Comment by przepompownia** [comment]
I'll try to catch it in my free time. I don't get this failure very often and sometimes it's not easy to remember what edition has been made right before crash :wink:

**Comment by b0o** [comment]
Same issue here. With Noice installed, it happens every time I do `:S<crashes immediately>`

**Comment by przepompownia** [comment]
I haven't had any such crash with Noice so far and cannot reproduce it that way. Maybe configuration matters here.

**Comment by przepompownia** [comment]
```lua
local function setupNoice()
  require('noice').setup({
    messages = {
      view = 'mini',
      view_search = false,
      view_error = 'mini',
      view_warn = 'mini',
    },
    notify = {
      enabled = true,
      view = 'mini',
    },
    lsp = {
      override = {},
      signature = {
        enabled = false,
        auto_open = {
          enabled = true,
          trigger = true,
          luasnip = true,
          throttle = 50,
        },
        view = nil, 
      hover = {
        enabled = false,
        view = nil,
        opts = {},
      },
    },
    views = {
      mini = {
        timeout = 5000,
      },
    },
    presets = {
      lsp_doc_border = true,
    },
  })
end
```

**Comment by luukvbaal** [comment]
> Same issue here. With Noice installed, it happens every time I do `:S<crashes immediately>`

Is this still true? If so, are you able to produce a minimal config that shows the issue? I've tried to reproduce with noice but couldn't.

And @przepompownia have you hit the assert after #27047?

**Comment by przepompownia** [comment]
Since #27047 I haven't this crash anymore. It's good moment to close this issue. Thanks @luukvbaal

**Comment by wookayin** [comment]
I was just hit by the same crash (same stacktrace) again (when doing undo). NVIM v0.10.0-dev-2139+g0c7ad223f; the equivalent base commit is fa9a85ae4.

**Comment by luukvbaal** [comment]
Found a semi-reliable reproducer, looking into it...

**Comment by Armadillidiid** [comment]
Same error over here

**Comment by przepompownia** [comment]
Today I got this failure on master before merging #27122. Let's observe nvim with that fix for some time.

**Comment by bassamsdata** [comment]
Hi, I'm encountering the issue with the latest nightly release (video below). I attempted to reproduce it in my configuration, but it seems to be quite random. However, I experimented with LazyVim distro and found a pattern that can almost reproduce it.

I'm still working on minimizing the reproducible code, and I'll likely post it here tomorrow.

```sh
NVIM v0.10.0-dev-2397+ga9f578b7a
Build type: RelWithDebInfo
LuaJIT 2.1.1707061634
```

https://github.com/neovim/neovim/assets/105807570/c65a752b-b60a-4596-a93c-de1ffa60178a

**Comment by bassamsdata** [comment]
Alright, here is the reproducible code for this issue:

Steps to reproduce it:

1. Execute the following commands in your terminal:
```sh
mkdir ~/.config/min_sign
cd ~/.config/min_sign
touch init.lua
```

2. Add the following code to the `init.lua` file then `NVIM_APPNAME=min_sign nvim`:

<details>

```lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
	vim.fn.system({
		"git",
		"clone",
		"--filter=blob:none",
		"https://github.com/folke/lazy.nvim.git",
		"--branch=stable", -- latest stable release
		lazypath,
	})
end
vim.opt.rtp:prepend(lazypath)


-- this is from LazyVim distro (simplified version similar to statuscol plugin)
-- https://github.com/LazyVim/LazyVim/blob/a50f92f7550fb6e9f21c0852e6cb190e6fcd50f5/lua/lazyvim/util/ui.lua
local function get_signs(buf, lnum)
	-- Get regular signs
	local signs = {}

	if vim.fn.has("nvim-0.10") == 0 then
		-- Only needed for Neovim <0.10
		-- Newer versions include legacy signs in nvim_buf_get_extmarks
		for _, sign in ipairs(vim.fn.sign_getplaced(buf, { group = "*", lnum = lnum })[1].signs) do
			local ret = vim.fn.sign_getdefined(sign.name)[1]
			if ret then
				ret.priority = sign.priority
				signs[#signs + 1] = ret
			end
		end
	end

	-- Get extmark signs
	local extmarks = vim.api.nvim_buf_get_extmarks(
		buf,
		-1,
		{ lnum - 1, 0 },
		{ lnum - 1, -1 },
		{ details = true, type = "sign" }
	)
	for _, extmark in pairs(extmarks) do
		signs[#signs + 1] = {
			name = extmark[4].sign_hl_group or "",
			text = extmark[4].sign_text,
			texthl = extmark[4].sign_hl_group,
			priority = extmark[4].priority,
		}
	end

	-- Sort by priority
	table.sort(signs, function(a, b)
		return (a.priority or 0) < (b.priority or 0)
	end)

	return signs
end

local function get_mark(buf, lnum)
	local marks = vim.fn.getmarklist(buf)
	vim.list_extend(marks, vim.fn.getmarklist())
	for _, mark in ipairs(marks) do
		if mark.pos[1] == buf and mark.pos[2] == lnum and mark.mark:match("[a-zA-Z]") then
			return { text = mark.mark:sub(2), texthl = "DiagnosticHint" }
		end
	end
end

local function icon(sign, len)
	sign = sign or {}
	len = len or 2
	local text = vim.fn.strcharpart(sign.text or "", 0, len) ---@type string
	text = text .. string.rep(" ", len - vim.fn.strchars(text))
	return sign.texthl and ("%#" .. sign.texthl .. "#" .. text .. "%*") or text
end

local function foldtext()
	local start_line = vim.api.nvim_buf_get_lines(0, vim.v.foldstart - 1, vim.v.foldstart, false)[1]
	local end_line = vim.api.nvim_buf_get_lines(0, vim.v.foldend - 1, vim.v.foldend, false)[1]

	-- Replace tabs with spaces
	start_line = start_line:gsub("\t", string.rep(" ", vim.bo.tabstop))

	-- Trim whitespace
	end_line = end_line:gsub("^%s*(.-)%s*$", "%1")

	return start_line .. " ... " .. end_line
end

function Statuscolumn()
	local win = vim.g.statusline_winid
	local buf = vim.api.nvim_win_get_buf(win)
	local is_file = vim.bo[buf].buftype == ""
	local show_signs = vim.wo[win].signcolumn ~= "no"

	local components = { "", "", "" } -- left, middle, right

	if show_signs then
		-- -@type Sign?,Sign?,Sign?
		local left, right, fold
		for _, s in ipairs(get_signs(buf, vim.v.lnum)) do
			if s.name and s.name:find("GitSign") then
				right = s
			else
				left = s
			end
		end
		if vim.v.virtnum ~= 0 then
			left = nil
		end
		vim.api.nvim_win_call(win, function()
			if vim.fn.foldclosed(vim.v.lnum) >= 0 then
				fold = {
					text = vim.opt.fillchars:get().foldclose or "",
					texthl = "Folded",
				}
			end
		end)
		-- Left: mark or non-git sign
		components[1] = icon(get_mark(buf, vim.v.lnum) or left)
		-- Right: fold icon or git sign (only if file)
		components[3] = is_file and icon(fold or right) or ""
	end

	-- Numbers in Neovim are weird
	-- They show when either number or relativenumber is true
	local is_num = vim.wo[win].number
	local is_relnum = vim.wo[win].relativenumber
	if (is_num or is_relnum) and vim.v.virtnum == 0 then
		if vim.v.relnum == 0 then
			components[2] = is_num and "%l" or "%r" -- the current line
		else
			components[2] = is_relnum and "%r" or "%l" -- other lines
		end
		components[2] = "%=" .. components[2] .. " " -- right align
	end

	return table.concat(components, "")
end

-- TODO: this is important
vim.opt.statuscolumn = [[%!v:lua.Statuscolumn()]]

-- we need mini.hipatterns to add icons as sign text
require("lazy").setup({
	{
		"echasnovski/mini.hipatterns",
		config = function()
			local mini_hipatterns = require("mini.hipatterns")
			if mini_hipatterns then
				local keywords = { "NOTE", "BUG", "LOVE", "TODO", "TEST", "FIX" }
				local icons = { "󱑴", "󱑬", "󱑭", "󱑯", "󱑱" }
				local highlighters = {}
				for _, keyword in ipairs(keywords) do
					local lowerKeyword = string.lower(keyword)
					highlighters[lowerKeyword] = {
						pattern = string.format("%s:", keyword),
						group = "IncSearch",
						extmark_opts = {
							sign_text = icons[_],
							sign_hl_group = "IncSearch",
						},
					}
					highlighters[string.format("%s_trail", lowerKeyword)] = {
						pattern = string.format("%s: ()%%S+.*()", keyword),
						group = "IncSearch",
					}
				end

				mini_hipatterns.setup({ highlighters = highlighters })
			end
		end,
	},
}, {})

vim.wo.number = true
vim.o.undofile = true
vim.wo.signcolumn = "yes"
vim.o.termguicolors = true

```
</details>

4. Edit the file `:e init.lua`.
5. Enter insert mode and write multiple TODOs in different lines (around 5 or six with multiple lines).
6. Visual select all those TODOs and delete them all at once.
7. The issue should occur.

Note1: If the issue doesn't occur, try writing multiple far away TODOs and then visually selecting and deleting them.

Note2: If the highlight doesn't have an sign text icon like `FIX:`, the issue won't happen. For example, writing `FIX: sometext` which doesn't have a sign text icon will not trigger the issue.

Note3: The code section may seem lengthy, but it comprises only two main parts:
One: The LazyVim distro's minimal/simple implementation of statuscol, resembling the statuscol plugin.
Second: Mini.hipatterns plugin, responsible for adding sign text icons and highlights.

Note4: in my own configuration, it doesn't happen like this, but it was so random, when sometimes I try to comment out some highlighted text that has icon sign text.

I apologize for the non-perfect minimal repro, considering the randomness of this issue, this is the best I could do.

https://github.com/neovim/neovim/assets/105807570/63d2c647-261b-4343-aa67-d39cf3ce4027


(video shows that if only a couple of TODOs are deleted, the issue won't occur, but if multiple are deleted, it will happen).

Let me know if you need further clarification.

**Comment by luukvbaal** [comment]
I was able to reproduce. In the repro, signs are counted multiple times in the overlap/non-overlap iteration in `buf_signcols_count_range()`:
https://github.com/neovim/neovim/blob/9bb046d1be5aa9ba0482b2cad050b286d4b78978/src/nvim/decoration.c#L800-L824

I'm not sure if this is a bug in the marktree API or if its used incorrectly @bfredl.

**Comment by luukvbaal** [comment]
Here is a test case which shows an unexpected amount of signs in the screen state and crashes upon deleting the first sign.

```lua
  it('correct number of signs after deleting text(#27046)', function()
    command('call setline(1, ["foo bar"]->repeat(31))')
    api.nvim_buf_set_extmark(0, ns, 0, 0, {end_row = 0, sign_text = 'S1'})
    api.nvim_buf_set_extmark(0, ns, 0, 0, {end_row = 0, end_col = 3, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 0, 4, {end_row = 0, end_col = 7, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 9, 0, {end_row = 9,  sign_text = 'S2'})
    api.nvim_buf_set_extmark(0, ns, 9, 0, {end_row = 9, end_col = 3, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 9, 4, {end_row = 9, end_col = 7, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 19, 0, {end_row = 19, sign_text = 'S3'})
    api.nvim_buf_set_extmark(0, ns, 19, 0, {end_row = 19, end_col = 3, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 19, 4, {end_row = 19, end_col = 7, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 29, 0, {end_row = 29, sign_text = 'S4'})
    api.nvim_buf_set_extmark(0, ns, 29, 0, {end_row = 29, end_col = 3, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 29, 4, {end_row = 29, end_col = 7, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 30, 0, {end_row = 30, sign_text = 'S5'})
    api.nvim_buf_set_extmark(0, ns, 30, 0, {end_row = 30, end_col = 3, hl_group = 'Error'})
    api.nvim_buf_set_extmark(0, ns, 30, 4, {end_row = 30, end_col = 7, hl_group = 'Error'})
    command('0d29')

    screen:expect{grid=[[
      S1S2S3S4S5{4:^foo} {4:bar}                         |
      S5{1:              }{4:foo} {4:bar}                           |
      29 fewer lines                                    |
    ]]}
    api.nvim_buf_clear_namespace(0, ns, 0, -1)
  end)
```

**Comment by przepompownia** [comment]
Thanks to all of you for investigating what happened here.

**Comment by stagnation** [comment]
I have a source file that crashes today. Immediately when opening it as checked in, or if I delete all the content and add new lines it crashes after a while.

> neovim-0.10.0~ubuntu1+git202404081538-978962f9a-971e32c87
Using statuscol with the 0.10 branch, if I copy the file it is fine so I think there are cached signs involved for that filepath.
Everything is fine without statuscol - but then I do not have the nice gitsigns.

![image](https://github.com/neovim/neovim/assets/1367522/683ab170-426e-4c75-8557-00976ea0078b)

Here are some of the marks if that helps. I am a little foggy on how best to debug signs/extmarks so let me know what to run.

<details>
<summary>
`lua print(vim.inspect(vim.fn.sign_getplaced(1, {group='*'})))`
</summary>

```
{ {
    bufnr = 1,
    signs = { {
        group = "gitsigns_extmark_signs_",
        id = 1,
        lnum = 1,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 2,
        lnum = 2,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 4,
        lnum = 4,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 5,
        lnum = 5,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 6,
        lnum = 6,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 8,
        lnum = 8,
        name = "",
        priority = 6
      }, {
        group = "gitsigns_extmark_signs_",
        id = 9,
        lnum = 9,
        name = "",
        priority = 6
      }, {
        group = "",
        id = 104001,
        lnum = 104,
        name = "Signature_j",
        priority = 10
      }, {
        group = "",
        id = 96001,
        lnum = 104,
        name = "Signature_j",
        priority = 10
      } }
  } }

```
</details>

**Comment by luukvbaal** [comment]
> Here are some of the marks if that helps.

Thanks, but not not really. What matters here is how the signs are placed/removed. Really need an exact reproducer at this point (with plugins is fine).

**Comment by stagnation** [comment]
Thanks. I'll see what I can do, is there any serialized state to be aware of? I see that as I add more and more lines I get more and more signs, so they seem to come from online parsing of git in this case.

**Comment by luukvbaal** [comment]
Thanks for working on this! I'm not sure how shada is relevant sorry, AFAIK it does not store extmarks :/ statuscol plugin should also not be needed to reproduce any crash, try `:set signcolumn=auto:2` instead.

Sign outside of buffer range might be a good clue, I'm not sure if that is handled properly.

**Comment by stagnation** [comment]
Yes! This reproduces and crashes immediately. Without updatetime it does not **seem to** crash 

```
 let plug_location = '~/.local/share/nvim/plugged'

 call plug#begin(plug_location)

Plug 'kshenoy/vim-signature', { 'commit': '6bc3dd1294a2' }

 call plug#end()

set updatetime=100
set noswapfile                 " Not required but makes life less annoying
set signcolumn=auto:2
```

**Comment by TamaMcGlinn** [comment]
@bfredl [your fix for this](https://github.com/neovim/neovim/commit/cb5ae22eab1fbb3162d0f7440a48ad49c7b3c48e) was merged in 0.10 already, but I have this crash consistently on neovim version 0.11:

```
$ nvim --version
NVIM v0.11.0-dev-164+g3a1515bfe
Build type: RelWithDebInfo
LuaJIT 2.1.1716656478
Run "nvim -V1 -v" for more info
```

It seems I can reproduce this consistently, by searching using this [ctrlsf](https://github.com/dyng/ctrlsf.vim) binding and then switching to the opened search window:

```
vmap <Leader>8 <Plug>CtrlSFVwordPath
```

But I've done these same steps before with the same NeoVim and config, without it crashing. Let me know soon if you want more info, so I can do that before it magically disappears.

**Comment by luukvbaal** [comment]
@TamaMcGlinn there have been multiple fixes after the one you linked. But this is not much to go on. If you can still reproduce on the latest commit, please open an issue with minimal reproduction steps (plugins are fine for this issue but please try to reduce to a minimal list of plugins necessary, a text searching plugin seems completely unrelated).

**Comment by TamaMcGlinn** [comment]
After updating to the latest nightly, I can still reproduce. Unfortunately, not just when I have that search plugin - I need to do some bisecting in my vimrc to find which part of the config is necessary. However, I do have 9 coredump files of the crash, if that helps.

**Comment by luukvbaal** [comment]
> However, I do have 9 coredump files of the crash, if that helps.

Not really. The assertion indicates an error in tracking placed signs. What matters is how those signs are placed/removed.

**Comment by TamaMcGlinn** [comment]
After building from source and manually bisecting, it seems that [perf(column): only invalidate lines affected by added sign](https://github.com/neovim/neovim/commit/f4001d27efae44c6c07678ad2c72eed5f1a25ea8) introduces the problem, as I can reliably reproduce the issue on that commit, but not on its parent (Merge pull request #26292 from luukvbaal/decor).

**Comment by TamaMcGlinn** [comment]
Ok, I'll open a new issue since the steps to reproduce are different. But the reason I was commenting is that the exact same assert is failing as in this issue, so I was thinking that maybe it also has the same cause as in this issue.

**Comment by luukvbaal** [comment]
> Sign outside of buffer range might be a good clue, I'm not sure if that is handled properly.

Found a way to still hit this assert with legacy signs with the hint from @stagnation (never looked into it then sorry). The `:sign` API still allows placing signs beyond the end of the buffer. The signcolumn tracking and possibly other marktree related code does not account for that.

Since they are now implemented as extmarks they can never actually become visible so maybe we should just error when trying to place signs beyond the end of the buffer like we with the extmark API @bfredl?

Unclear if this is causing the issue @TamaMcGlinn is facing.

**Comment by fm39hz** [comment]
![Image](https://github.com/user-attachments/assets/db938b38-1d56-4e90-af5d-ef65209788ed)
I have this issue again with save action
this is my nvim version
```
NVIM v0.12.0-dev-571+g7aafbca510
Build type: RelWithDebInfo
LuaJIT 2.1.1748459687
Run "nvim -V1 -v" for more info
```

**Comment by bfredl** [comment]
@fm39hz would you mind trying latest nightly? There was a change very recently to adjust the assert.

**Comment by fm39hz** [comment]
It seems like `v0.12.0-dev-630+g8af2aea24f` didn't have this issue anymore!
Glad that it has been fixed
